### PR TITLE
Improve attachment upload methods

### LINKED ISSUE #20130: Copy and paste the images has list legend
### Feature Description

At the moment, the pasted image in the issue content doesn't appear in the file list below (dropzone).

It's better to make the pasted image added into the file list legend below as the general upload file.

### Screenshots

image

### PR DESCRIPTION
**Pull request by tyroneyeh**
* Use dropzone to handle file uploading for all cases, including pasting and dragging
* Merge duplicate code, use consistent behavior for link generating

Close #20130

### CODE REVIEW

File: web_src/js/features/comp/ComboMarkdownEditor.js [hunk 1]
```diff
@@ -315,3 +310,9 @@ export async function initComboMarkdownEditor(container, options = {}) {
   await editor.init();
   return editor;
 }
+
+export function removeLinksInTextarea(editor, file) {
+  const fileName = file.name.slice(0, file.name.lastIndexOf('.'));
+  const fileText = `\\[${fileName}\\]\\(/attachments/${file.uuid}\\)`;
+  editor.value(editor.value().replace(new RegExp(`<img [\\s\\w"=]+ alt="${fileName}" src="/attachments/${file.uuid}">`, 'g'), '').replace(new RegExp(`\\!${fileText}`, 'g'), '').replace(new RegExp(fileText, 'g'), ''));
+}
```
Comment (silverwind) [inline web_src/js/features/comp/ComboMarkdownEditor.js]: Use https://github.com/sindresorhus/escape-string-regexp here, later to be replaced with https://github.com/tc39/proposal-regex-escaping.

Also, this function should have some unit tests.

File: web_src/js/features/comp/Paste.js [hunk 2]
```diff
@@ -82,35 +82,48 @@ class CodeMirrorEditor {
...
+      text = `[${name}](${url})`;
     }
     editor.replacePlaceholder(placeholder, text);
 
+    file.uuid = uuid;
+    dropzone.dropzone.emit('addedfile', file);
+    if (file.type?.startsWith('image/')) {
+      const imgSrc = `/attachments/${file.uuid}`;
+      dropzone.dropzone.emit('thumbnail', file, imgSrc);
+      dropzone.querySelector(`img[src='${imgSrc}']`).style.maxWidth = '100%';
```
Comment (silverwind) [inline web_src/js/features/comp/Paste.js]:
```suggestion
      dropzone.querySelector(`img[src='${CSS.escape(imgSrc)}']`).style.maxWidth = '100%';
```

File: web_src/js/features/repo-issue-edit.js [hunk 3]
```diff
@@ -78,16 +72,16 @@ async function onEditContent(event) {
...
             dropzone.querySelector('.files').innerHTML = '';
             for (const el of dropzone.querySelectorAll('.dz-preview')) el.remove();
-            fileUuidDict = {};
             disableRemovedfileEvent = false;
 
             for (const attachment of data) {
-              const imgSrc = `${dropzone.getAttribute('data-link-url')}/${attachment.uuid}`;
               dz.emit('addedfile', attachment);
-              dz.emit('thumbnail', attachment, imgSrc);
+              if (/\.(jpg|jpeg|png|gif|bmp|svg)$/i.test(attachment.name)) {
```
Comment (silverwind) [inline web_src/js/features/repo-issue-edit.js]: Try to get the mime type into here and use it instead.

File: web_src/js/utils/dom.js [hunk 4]
```diff
@@ -258,16 +258,27 @@ export function isElemVisible(element) {
   return Boolean(element.offsetWidth || element.offsetHeight || element.getClientRects().length);
 }
 
+export function getComboMarkdownEditor(el) {
+  if (el?.jquery) el = el[0];
```
Comment (silverwind) [inline web_src/js/utils/dom.js]: What is `.jquery`?

File: web_src/js/utils/dom.js [hunk 5]
```diff
@@ -258,16 +258,27 @@ export function isElemVisible(element) {
...
-  for (const item of e.clipboardData?.items ?? []) {
-    if (item.type?.startsWith('image/')) {
-      images.push(item.getAsFile());
+  const acceptedFiles = getComboMarkdownEditor(e.currentTarget).dropzone.getAttribute('data-accepts');
+  const files = [];
+  const data = e.clipboardData?.items || e.dataTransfer?.items;
+  for (const item of data ?? []) {
+    if (item?.kind === 'file') {
+      const file = item.getAsFile();
+      const extName = file.name.slice(file.name.lastIndexOf('.'), file.name.length);
```
Comment (silverwind) [inline web_src/js/utils/dom.js]: Use https://github.com/go-gitea/gitea/blob/02e183bf3fa502b7cef76e8dcdbf01b85ce641f0/web_src/js/utils.js#L10

File: web_src/js/utils/dom.js (hunk 4 above)
Comment (tyroneyeh) [reply web_src/js/utils/dom.js]: `.jquery` displays the jquery version, or null if el is not a jquery element

File: web_src/js/features/repo-issue-edit.js (hunk 3 above)
Comment (tyroneyeh) [reply web_src/js/features/repo-issue-edit.js]: The attachment variable has no type attribute. This information is obtained from the database. The database does not record the file type.

Comment (silverwind) [reply web_src/js/features/repo-issue-edit.js]: There is a typesniffer package in the backend that returns the mime type based on filename. Possible to use here?

https://github.com/go-gitea/gitea/blob/d1ba5f431b09cf2a1f814ce0b3feb4c1d4755844/modules/typesniffer/typesniffer.go#L85

File: web_src/js/utils/dom.js (hunk 4 above)
Comment (silverwind) [reply web_src/js/utils/dom.js]: Use `el instanceof $` instead. Also see https://stackoverflow.com/questions/1853223.

Comment (tyroneyeh) [reply web_src/js/utils/dom.js]: If jquery is not imported, this writing method will cause an error.

Comment (silverwind) [reply web_src/js/utils/dom.js]: Could just import it above, but seeing that we are in the middle of removing jQuery, I would suggest this function to only accept `HTMLElement`.

Comment (tyroneyeh) [reply web_src/js/utils/dom.js]: solved

**Comment by wxiaoguang** [comment]
Will take a look later (resolve the conflicts and do some improvements)

File: web_src/js/features/comp/ComboMarkdownEditor.js (hunk 1 above)
Comment (wxiaoguang) [reply web_src/js/features/comp/ComboMarkdownEditor.js]: IMO this function is not needed, why it should replace HTML `` syntax to markdown syntax.

File: web_src/js/utils/dom.js [hunk 6]
```diff
@@ -258,16 +259,25 @@ export function isElemVisible(element) {
   return Boolean(element.offsetWidth || element.offsetHeight || element.getClientRects().length);
 }
 
+export function getComboMarkdownEditor(el) {
+  return el?._giteaComboMarkdownEditor;
+}
+
```
Comment (silverwind) [inline web_src/js/utils/dom.js]: Move this somewhere else, `dom.js` is for general DOM-related stuff and this is too specific to the markdown editor.

**Comment by wxiaoguang** [comment]
Still WIP, will fill some code back later.

**Comment by wxiaoguang** [comment]
->  Refactor dropzone #31482 

I think we need to refactor the legacy code first, have a clear codebas and then continue. Otherwise there are too many unclear behaviors

**Comment by wxiaoguang** [comment]
It works again now (while there are still some TODO/FIXME problems)

**Comment by wxiaoguang** [comment]
Almost done and ready for review.

File: web_src/js/utils.js [hunk 7]
```diff
@@ -142,3 +144,7 @@ export function serializeXml(node) {
 }
 
 export const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
+
+export function isWellKnownImageFilename(fn) {
+  return /\.(jpe?g|png|gif|webp|svg)$/i.test(fn);
```
Comment (silverwind) [inline web_src/js/utils.js]:
```suggestion
  return /\.(jpe?g|png|gif|webp|svg|heic)$/i.test(fn);
```

Somewhat common in Apple ecosystem.

**Comment by wxiaoguang** [comment]
All done IMO

**Review by wxiaoguang — APPROVED** [review]

File: web_src/js/features/comp/EditorUpload.js [hunk 8]
```diff
@@ -1,19 +1,29 @@
...
+import {
+  DropzoneCustomEventRemovedFile,
+  DropzoneCustomEventUploadDone,
+  generateMarkdownLinkForAttachment,
+} from '../dropzone.js';
+
+let uploadIdCounter = 0;
+
+function uploadFile(dropzoneEl, file) {
+  return new Promise((resolve, _) => {
```
Comment (silverwind) [inline web_src/js/features/comp/EditorUpload.js]:
```suggestion
  return new Promise((resolve) => {
```

**Review by silverwind — APPROVED** [review]

**Pull request merged**